### ISSUE TITLE
A core dump error happened in gearman_worker_free at the latest 1.1.20 gearmand

### ISSUE THREAD

**Issue by mrbejka** [gearman/gearmand#368]
Hi,

        In my use, I find a segment error which is the same as #155. And I try to upgrade to 1.1.20, this will not happen agian but we get a new segment error, the core dump stack can be seems as below：
<img width="778" alt="gearman error" src="https://github.com/gearman/gearmand/assets/33799358/97954109-dacd-4e0d-9bac-7a4e3f0a16f7">
<img width="786" alt="debug gearmand" src="https://github.com/gearman/gearmand/assets/33799358/ae412ba6-ec9d-4ed3-91d6-955a75934946">

        After I add reset_recv_packet , my program will not report segfault with gearman .  

<img width="747" alt="resolve gearmand" src="https://github.com/gearman/gearmand/assets/33799358/7599105e-bb91-4632-be78-17d111eb9c6d">

       I don't know if my modification is a robust way to resolve this problem as I know very little about the actual mechanism of gearman. This seems that _recv_packet is ofen set to the address of Job::asssigned, and connection.cc should have no permission to free __recv_packet as this __recv_packet is always borrowed from Job in my option.

**Comment by esabol** [comment]
Screenshots are not the best way to report issues or suggest changes in my opinion.

To clarify, you added `reset_recv_packet();` after line 1037 here:

https://github.com/gearman/gearmand/blob/442606718983ffb2b1da7820cc6698cf56d01959/libgearman/connection.cc#L1033-L1040

I can't see any harm in doing so, but I'm not sure why it would be needed.

If this does solve a problem, I wonder if it might also be necessary before line 1080:

https://github.com/gearman/gearmand/blob/442606718983ffb2b1da7820cc6698cf56d01959/libgearman/connection.cc#L1078-L1082

Any thoughts on this, @SpamapS ?

**Comment by mrbejka** [comment]
Sorry for report a issue in a unpleasant way, this is my first issue and first contact with github developers.

This is my  flow in my program, it initialize a gearman_worker_st called worker at first, and it would finalize and free this worker at the end. Between initialize and finalize, my program would call gearman_worker_work with timeout seted and check the ret of gearman_worker_free, when gearman_worker_work return GEARMAN_NOT_CONNECTED or GEARMAN_COULD_NOT_CONNECT, it would try to call gearman_worker_free to finalize the older gearman_worker_st and initialize a new gearman_worker_st, and the program would print some log before calling gearman_worker_free.  

After I upgrade gearman to 1.1.20, I would not get segfault at gearman_worker_work,but I would get new segfault at gearman_worker_free. I don't know why gearman_worker_free was called because I didn't find any logs about the return, but my program does call gearman_worker_free and causes segfault. And the coredump info can be seen as the preivous screenshots 1-2. 

As the screenshot shows, the _recv_packet is not a regular gearman_packet_st packet, so I thought this may be caused by this __recv_packet is borrowed from the foreign object and reset_recv_packet is not called after recv_state is set to GEARMAN_CON_RECV_UNIVERSAL_NONE.

Do these words help you understand what's happening better now?

**Comment by esabol** [comment]
Yes, that was very helpful. Thank you for the additional information, @mrbejka.

**Comment by SpamapS** [comment]
I'm not super familiar with this code or what is happening but it would be
helpful if you could provide us with a very short example bit of code that
would reproduce the issue reliably. Please also include your OS details.
Thanks!

**Comment by esabol** [comment]
@mrbejka , can you provide us with worker code which reproduces the core dump? It would really help!

**Comment by mrbejka** [comment]
Hi，easol and SpampS

     I am pleased to offer you a case,but I'm sorry that I can't not offer the origin case to you. But I would try to design a new case to help reproduce it. This need times as I'm busy. 

    This happen at CentOS7 , and blocking api of gearman are used only in my case. In my option,the key points that lead to this phenomenon are the large number of worker threads, the occasional big data transfer and the blocking api.

**Comment by esabol** [comment]
Reviewing old issues.... This issue was _probably_ fixed in the 2.1.0 release. I think it can be closed. @p-alik : Do you agree?

**Comment by p-alik** [comment]
I don't think this was fixed in 2.1.0. The code paths involved are unchanged from the blob referenced above (`4426067`) through the `2.1.0` tag and current `master`.

The only vaguely related change since then is `3c13f803` (part of the #413 fix), which makes `recv_socket()` call `close_socket()` on `GEARMAN_SHUTDOWN_GRACEFUL` as well as `GEARMAN_SHUTDOWN`. That addresses one narrow dangling-`_recv_packet` case tied to the internal wakeup-pipe shutdown, but not the network-error paths this issue is about.

@mrbejka's root-cause analysis still holds. In `gearman_connection_st::receiving()`, `_recv_packet` is aliased to the caller's packet (`gearman_packet_create()` just returns `&packet_arg`; from the worker path that's `worker->job()->impl()->assigned`). The failure path:

1. Blocking `gearman_wait()` inside `recv_socket()` returns a failure other than `GEARMAN_SHUTDOWN[_GRACEFUL]` (e.g. `GEARMAN_TIMEOUT`) → no `close_socket()` (`connection.cc:1245-1253`).
2. `receiving()` sets `recv_state = GEARMAN_CON_RECV_UNIVERSAL_NONE` but returns without `reset_recv_packet()` (`connection.cc:1050-1056`), so `_recv_packet` still points at `assigned`.
3. Worker calls `worker->job(NULL)` → `gearman_job_free()` → `delete job` (`worker.cc:866`), freeing that memory.
4. `_recv_packet` now dangles. The next `close_socket()` with a live fd — on reconnect, and again during `gearman_worker_free` teardown — calls `gearman_packet_free()` on freed memory → segfault.

That matches the reported crash site and the coredump ("`_recv_packet` is not a regular `gearman_packet_st`"). The same shape exists at `connection.cc:1037` and `:1080`, which are only safe when `close_socket()` runs with a valid fd.

Minimal fix is still what was proposed originally: reset the borrowed `_recv_packet` on every non-`IO_WAIT` error exit of `receiving()` (or route those exits through `close_socket()`). I'd keep this open.

**Comment by esabol** [comment]
Well, thanks to @p-alik we finally have a fix for this issue, so the issue has been closed.

@p-alik : I'm a little reluctant to release a new version with just this fix. I'd like to have a couple more things in the release notes. If there are any other issues that you want to work on, let me know....
